File: SEGUNDO PERIODO/AEDS II/ESTUDOS/Pilha Flexivel/Pilha.java
class No {
    int elemento;
    No prox;

    No(int elemento) {
        this.elemento = elemento;
        this.prox = null;
    }
}

public class Pilha {
    private No topo;

    // Construtor
    Pilha() {
        topo = null;
    }

    // Inserir elemento no topo
    public void inserir(int x) {
        No novo = new No(x);
        novo.prox = topo;
        topo = novo;
    }

    // Remover elemento
    public int remover() {
        if (topo == null) {
            System.out.println("Erro: pilha vazia!");
            System.exit(1);
        }
        int elemento = topo.elemento;
        No tmp = topo;
        topo = topo.prox;
        tmp.prox = null;
        tmp = null;
        return elemento;
    }

    // Imprimir reverso
    private void imprimirReverso(No i) {
        if (i != null) {
            imprimirReverso(i.prox);
            System.out.print(i.elemento + " ");
        }
    }

    // Mostrar pilha
    public void mostrar() {
        imprimirReverso(topo);
        System.out.println();
    }

    public static void main(String[] args) {
        Pilha p = new Pilha();

        p.inserir(3);
        p.inserir(5);
        p.inserir(7);

        System.out.print("Pilha antes da remoção: ");
        p.mostrar();

        int removido = p.remover();
        System.out.println("Elemento removido: " + removido);

        System.out.print("Pilha depois da remoção: ");
        p.mostrar();
    }
}

File: SEGUNDO PERIODO/AEDS II/ESTUDOS/Pilha Flexivel/Pilha.c
#include <stdio.h>
#include <stdlib.h>

// Estrutura de nó
typedef struct No
{
    int elemento;
    struct No *prox;
} No;

// Estrutura da pilha
typedef struct Pilha
{
    No *topo;
} Pilha;

// Cria uma nova pilha
Pilha *newPilha()
{
    Pilha *tmp = (Pilha *)malloc(sizeof(Pilha));
    tmp->topo = NULL;
    return tmp;
}

// Cria um novo nó
No *newNo(int x)
{
    No *tmp = (No *)malloc(sizeof(No));
    tmp->elemento = x;
    tmp->prox = NULL;
    return tmp;
}

// Insere elemento na pilha
void inserirPilha(int x, Pilha *p)
{
    No *novo = newNo(x);
    novo->prox = p->topo;
    p->topo = novo;
}

// Remove o elemento do topo da pilha
int removerPilha(Pilha *p)
{
    if (p->topo == NULL)
    {
        printf("Erro: pilha vazia!\n");
        exit(1);
    }

    int elemento = p->topo->elemento;
    No *tmp = p->topo;
    p->topo = p->topo->prox;
    tmp->prox = NULL;
    free(tmp);
    tmp = NULL;

    return elemento;
}

// Imprime a pilha de forma reversa (base → topo)
void imprimirReverso(No *topo)
{
    if (topo != NULL)
    {
        imprimirReverso(topo->prox);
        printf("%d ", topo->elemento);
    }
}

// Programa principal
int main()
{
    Pilha *p = newPilha();

    inserirPilha(3, p);
    inserirPilha(5, p);
    inserirPilha(7, p);

    printf("Pilha antes da remoção: ");
    imprimirReverso(p->topo);
    printf("\n");

    int removido = removerPilha(p);
    printf("Elemento removido: %d\n", removido);

    printf("Pilha depois da remoção: ");
    imprimirReverso(p->topo);
    printf("\n");

    return 0;
}
